hal/src/nRF52840/usart_hal.cpp

```diff
@@ -265,38 +260,41 @@ class Usart {
     }
 
     int suspend() {
-        CHECK_TRUE(isEnabled(), SYSTEM_ERROR_NONE);
-        CHECK_FALSE(suspend_, SYSTEM_ERROR_NONE);
+        CHECK_TRUE(isEnabled(), SYSTEM_ERROR_INVALID_STATE);
+
+        flush();
 
         AtomicSection lk;
 
-        flush();
+        // Update current available received data
+        data();
+        
         CHECK(disable());
 
         // Configuring the input pins' mode to PIN_MODE_NONE will disconnect input buffer to enable power savings
         // Configuring the output pins' mode to INPUT_PULLUP to not polluting the state on other side.
         // NOTE: Do not reset pin function!
         HAL_Pin_Mode(txPin_, INPUT_PULLUP);
         HAL_Pin_Mode(rxPin_, PIN_MODE_NONE);
-        if (config_.config & SERIAL_FLOW_CONTROL_CTS) {
-            HAL_Pin_Mode(ctsPin_, PIN_MODE_NONE);
-        }
         if (config_.config & SERIAL_FLOW_CONTROL_RTS) {
             HAL_Pin_Mode(rtsPin_, INPUT_PULLUP);
         }
+        if (config_.config & SERIAL_FLOW_CONTROL_CTS) {
+            HAL_Pin_Mode(ctsPin_, PIN_MODE_NONE);
+        }
 
-        suspend_ = true;
-        enabled_ = false;
+        state_ = HAL_USART_STATE_SUSPENDED;
         transmitting_ = false;
         receiving_ = 0;
-        rxBuffer_.reset();
+        rxBuffer_.prune();
```

Comment:
This should potentially not be needed since we've ran `data()` previously.

Reply:
Without clearing the pending head and tail, the available data is incorrect after restoring the UART. The `user/tests/wiring/serial_loopback2` will fail.

Reply:
I'm saying that they should not be cleared but commited (as in `data()`) function.

Reply:
We should at least clear the `headPending_`, otherwise, when restoring the UART the receiver cannot be started, as the acquirable size will be 0. The value of `headPending_` after calling `data()` is actually the aborted acquired space of the RX ring buffer.